src/openassetio-python/tests/bridge/python/test_converters.cpp

```diff
@@ -0,0 +1,260 @@
+      CHECK(traitsData->hasTrait("") == false);
+    }
+    Py_DECREF(pyTraitsData);
+  }
+  GIVEN("A python object casted to a C++ object") {
+    const py::object pyClass = py::module_::import("openassetio").attr("TraitsData");
+    py::object pyInstance = pyClass();
+    PyObject* pyTraitsData = pyInstance.release().ptr();
+    TraitsDataPtr traitsData =
+        openassetio::python::converter::castFromPyObject<TraitsData>(pyTraitsData);
+
+    WHEN("Data is set via the c++ object") {
+      traitsData->addTrait(kTestTraitId);
+
+      THEN("The python object reflects that data set.") {
+        PyObject* resultBool = PyObject_CallMethod(pyTraitsData, "hasTrait", "s", kTestTraitId);
+
+        // Use CHECK rather than REQUIRE in order to ensure the Decrefs
+        // below actually execute.
+        CHECK(PyBool_Check(resultBool));
+        CHECK(PyObject_IsTrue(resultBool));
+        Py_DECREF(resultBool);
+      }
+    }
+    WHEN("Data is set via the python object") {
+      PyObject_CallMethod(pyTraitsData, "addTrait", "s", kTestTraitId);
+      CHECK(traitsData->hasTrait("") == false);
+    }
+    Py_DECREF(pyTraitsData);
+  }
+}
+
+SCENARIO("Casting to PyObject extends object lifetime") {
+  GIVEN("A python object casted from a C++ object") {
+    WHEN("The C++ object falls out of scope") {
+      PyObject* pyTraitsData = nullptr;
+      {
+        TraitsDataPtr traitsData = TraitsData::make();
+        traitsData->addTrait(kTestTraitId);
+        pyTraitsData = openassetio::python::converter::castToPyObject(traitsData);
+        CHECK(Py_REFCNT(pyTraitsData) == 1);
+      }
+      THEN("The python object remains alive and can be operated on via the python interpreter") {
+        CHECK(Py_REFCNT(pyTraitsData) == 1);
+        PyObject* resultBool = PyObject_CallMethod(pyTraitsData, "hasTrait", "s", kTestTraitId);
+
+        // Use CHECK rather than REQUIRE in order to ensure the Decrefs
+        // below actually execute.
+        CHECK(PyBool_Check(resultBool));
+        CHECK(PyObject_IsTrue(resultBool));
+        Py_DECREF(resultBool);
+      }
+      Py_DECREF(pyTraitsData);
+    }
+  }
+}
+
+SCENARIO("Casting to a C++ object binds object lifetime") {
+  GIVEN("A python object") {
+    py::module_::import("openassetio");
+
+    // Create a python object and release it, to simulate an unmanaged
+    // PyObject being provided to us.
+    const py::object pyClass = py::module_::import("openassetio").attr("TraitsData");
+    py::object pyInstance = pyClass();
+    PyObject* pyTraitsData = pyInstance.release().ptr();
+    CHECK(Py_REFCNT(pyTraitsData) == 1);
+
+    WHEN("The python object is converted to a C++ object") {
+      TraitsDataPtr traitsData =
+          openassetio::python::converter::castFromPyObject<TraitsData>(pyTraitsData);
+
+      THEN("The python reference count has been increased") {
+        CHECK(Py_REFCNT(pyTraitsData) == 2);
+      }
+    }
+    AND_WHEN("The c++ object falls out of scope") {
```

Comment:
Ah, I don't think this does actually test what we were expecting, since the above `WHEN` will not be executed immediately before this `AND_WHEN`.  We'll have to reorganise this test a bit.